File: render_app/strategy.py
from collections import defaultdict
from datetime import datetime, timezone


def compute_pdh_pdl(bars: list[dict]) -> dict:
    pdh = max(b['high'] for b in bars)
    pdl = min(b['low'] for b in bars)
    return {'pdh': pdh, 'pdl': pdl, 'pdr': pdh - pdl}


def compute_vwap(bars: list[dict]) -> list[float]:
    result = []
    cum_tpv = 0.0
    cum_vol = 0
    for b in bars:
        tp = (b['high'] + b['low'] + b['close']) / 3
        cum_tpv += tp * b['volume']
        cum_vol += b['volume']
        result.append(cum_tpv / cum_vol if cum_vol > 0 else b['close'])
    return result


def compute_volume_avg(bars: list[dict], period: int = 20) -> list[float]:
    result = []
    for i in range(len(bars)):
        if i < period:
            result.append(0.0)
        else:
            total = sum(bars[j]['volume'] for j in range(i - period, i))
            result.append(total / period)
    return result


def check_signal(bar: dict, pdh: float, pdl: float, pdr: float,
                 vwap: float, vol_avg: float) -> dict | None:
    if vol_avg <= 0:
        return None
    if bar['high'] > pdh and bar['close'] > vwap and bar['volume'] >= 1.5 * vol_avg:
        return {
            'direction': 'BUY',
            'entry': pdh,
            'sl': pdh - 0.75 * pdr,
            'tp': pdh + 0.75 * pdr,
        }
    if bar['low'] < pdl and bar['close'] < vwap and bar['volume'] >= 1.5 * vol_avg:
        return {
            'direction': 'SELL',
            'entry': pdl,
            'sl': pdl + 0.75 * pdr,
            'tp': pdl - 0.75 * pdr,
        }
    return None


def zerodha_cost(qty: int, entry: float, exit: float) -> dict:
    turnover_buy = qty * entry
    turnover_sell = qty * exit
    total_to = turnover_buy + turnover_sell
    brokerage = min(0.0003 * turnover_buy, 20) + min(0.0003 * turnover_sell, 20)
    stt = 0.00025 * turnover_sell
    exchange_charges = 0.0000325 * total_to
    sebi = 0.000001 * total_to
    gst = 0.18 * (brokerage + exchange_charges)
    stamp_duty = 0.00015 * turnover_buy
    total = brokerage + stt + exchange_charges + sebi + gst + stamp_duty
    return {
        'brokerage': round(brokerage, 2),
        'stt': round(stt, 2),
        'exchange_charges': round(exchange_charges, 2),
        'sebi_charges': round(sebi, 2),
        'gst': round(gst, 2),
        'stamp_duty': round(stamp_duty, 2),
        'total': round(total, 2),
    }


def group_bars_by_date(bars: list[dict]) -> dict[str, list[dict]]:
    groups = defaultdict(list)
    for b in bars:
        key = b['date'].strftime('%Y-%m-%d')
        groups[key].append(b)
    return dict(groups)

File: lib/strategy.ts
export interface Bar {
  date: Date
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface PDHPDL {
  pdh: number
  pdl: number
  pdr: number
}

export interface SignalResult {
  direction: 'BUY' | 'SELL'
  entry: number
  sl: number
  tp: number
  barTime: Date
  pdh: number
  pdl: number
  pdr: number
  vwap: number
  volume: number
  volAvg: number
}

export interface CostBreakdown {
  brokerage: number
  stt: number
  exchangeCharges: number
  sebiCharges: number
  gst: number
  stampDuty: number
  total: number
}

export function computePDHPDL(bars: Bar[]): PDHPDL {
  let pdh = -Infinity
  let pdl = Infinity
  for (const b of bars) {
    if (b.high > pdh) pdh = b.high
    if (b.low < pdl) pdl = b.low
  }
  const pdr = pdh - pdl
  return { pdh, pdl, pdr }
}

export function computeVWAP(bars: Bar[]): number[] {
  const result: number[] = []
  let cumTPV = 0
  let cumVol = 0
  for (const b of bars) {
    const tp = (b.high + b.low + b.close) / 3
    cumTPV += tp * b.volume
    cumVol += b.volume
    result.push(cumVol > 0 ? cumTPV / cumVol : b.close)
  }
  return result
}

export function computeVolumeAvg(bars: Bar[], period: number = 20): number[] {
  const result: number[] = []
  for (let i = 0; i < bars.length; i++) {
    if (i < period) {
      result.push(0)
    } else {
      let sum = 0
      for (let j = i - period; j < i; j++) {
        sum += bars[j].volume
      }
      result.push(sum / period)
    }
  }
  return result
}

export function checkSignal(
  bar: Bar,
  pdh: number,
  pdl: number,
  pdr: number,
  vwap: number,
  volAvg: number,
): SignalResult | null {
  if (volAvg <= 0) return null

  if (bar.high > pdh && bar.close > vwap && bar.volume >= 1.5 * volAvg) {
    return {
      direction: 'BUY',
      entry: pdh,
      sl: pdh - 0.75 * pdr,
      tp: pdh + 0.75 * pdr,
      barTime: bar.date,
      pdh,
      pdl,
      pdr,
      vwap,
      volume: bar.volume,
      volAvg,
    }
  }

  if (bar.low < pdl && bar.close < vwap && bar.volume >= 1.5 * volAvg) {
    return {
      direction: 'SELL',
      entry: pdl,
      sl: pdl + 0.75 * pdr,
      tp: pdl - 0.75 * pdr,
      barTime: bar.date,
      pdh,
      pdl,
      pdr,
      vwap,
      volume: bar.volume,
      volAvg,
    }
  }

  return null
}

export function zerodhaCost(
  qty: number,
  entry: number,
  exit: number,
  _direction: 'BUY' | 'SELL',
): CostBreakdown {
  const turnoverBuy = qty * entry
  const turnoverSell = qty * exit
  const totalTurnover = turnoverBuy + turnoverSell

  const brokerageBuy = Math.min(0.0003 * turnoverBuy, 20)
  const brokerageSell = Math.min(0.0003 * turnoverSell, 20)
  const brokerage = brokerageBuy + brokerageSell

  const stt = 0.00025 * turnoverSell

  const exchangeCharges = 0.0000325 * totalTurnover

  const sebiCharges = 0.000001 * totalTurnover

  const gstValue = 0.18 * (brokerage + exchangeCharges)

  const stampDuty = 0.00015 * turnoverBuy

  const total = brokerage + stt + exchangeCharges + sebiCharges + gstValue + stampDuty

  return {
    brokerage,
    stt,
    exchangeCharges,
    sebiCharges,
    gst: gstValue,
    stampDuty,
    total,
  }
}

export function formatDateKey(d: Date): string {
  return d.toISOString().slice(0, 10)
}

export function groupBarsByDate(bars: Bar[]): Map<string, Bar[]> {
  const map = new Map<string, Bar[]>()
  for (const b of bars) {
    const key = formatDateKey(b.date)
    if (!map.has(key)) map.set(key, [])
    map.get(key)!.push(b)
  }
  return map
}
